library/serialize.c

```diff
@@ -0,0 +1,333 @@
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  This file is part of mbed TLS (https://tls.mbed.org)
+ */
+
+/*
+ * Serialization format description
+ *
+ * All integers are encoded as a stream of octets in big-endian format.
+ *
+ * A serialization packet consists of a 4-byte header followed by
+ * variable-sized content. The first byte of the header is the packet
+ * type:
+ *   0x45 ('E') execute   target->host  Execute a function
+ *   0x50 ('P') push      target->host  Push an input for the next execution
+ *   0x72 ('r') result    host->target  Return an output from the last execution
+ *
+ * The host maintains a stack of parameters and results. Stack
+ * elements are byte arrays. This stack is initially empty. The "push"
+ * command transmits one item from the target to the host and pushes
+ * this item to the stack. The "execute" command requests that the host
+ * perform an action, taking parameters from the stack and sending outputs
+ * back to the target.
+ *
+ * Format of a push packet (by byte offset):
+ *   0..0       0x50    Type identifier
+ *   1..3       size    Number of bytes to follow
+ *   4..3+size  data    Content of the byte array
+ *
+ * Format of an execute packet (by byte offset):
+ *   0..0       0x45    Type identifier
+ *   1..3       func    Function identifier
+ * Function identifiers are conventionally arranged as follows:
+ *   bits 23-16   Library code (1=sockets)
+ *   bits 15-8    Function identifier inside the library
+ *   bits 7-4     Number of parameters
+ *   bits 3-0     Number of results after status code on success
+ *
+ * Format of a result packet (by byte offset):
+ *   0..0       0x72    Type identifier
+ *   1..3       size    Number of bytes to follow
+ *   4..3+size  data    Content of the byte array
+ *
+ * All functions currently follow the following result status
+ * convention. The first result is a status code (MBEDTLS_ERR_xxx
+ * constant). If this is 0 then more results follow. If this is
+ * nonzero then it is the only result.
+ */
+
+#if !defined(MBEDTLS_CONFIG_FILE)
+#include "mbedtls/config.h"
+#else
+#include MBEDTLS_CONFIG_FILE
+#endif
+
+#include <limits.h>
+#include <stdint.h>
+#include <string.h>
+#include <signal.h>
+#include "mbedtls/serialize.h"
+
+#if defined(MBEDTLS_SERIALIZE_C)
+
+static int serialize_write_fd = -1;
+static int serialize_read_fd = -1;
+
+#if defined(MBEDTLS_SERIALIZE_FORK_FRONTEND_C)
```

Comment:
Because of it's platform specific nature I would prefer to move this code in a separate compilation unit, exemplifying one possible implementation of `mbedtls_serialize_write` and `mbedtls_serialize_read`. In this case, `mbedtls_serialize_read` and `mbedtls_serialize_write` would need to be unconditionally declared in `serialize.h`.